Refactor to always replace task details fragments instead of updating…

lemonboston opened the pull request:
… them when new task is selected. #624 

---

@dmfs It took some time to get everything correctly working here but it looks good now for me.

There is a field in `TaskListActivity` now to save the last used color, so it can be used while the selected task is loaded. It is useful upon starting up the app and after rotation.

`EmptyTaskFragment` also has a hint color now, so we don't change back the color to primary after a deletion but keep it as it was for that task. 

`ViewTaskFragment` can be further simplified after these changes, I've created a separate ticket for that: https://github.com/dmfs/opentasks/issues/628

This pull request should fix https://github.com/dmfs/opentasks/issues/609 and https://github.com/dmfs/opentasks/issues/525, too.

I've created 3 new classes for reading arguments from `Bundle` here, in order to read the color argument. They help now, but the plan would be to remove and replace them once the full solution from boxed-bolts is available, since that approach is much better and more general.

dmfs commented on opentasks/src/main/java/org/dmfs/tasks/TaskListFragment.java:
```diff
@@ -497,18 +506,13 @@ private void selectChildView(ExpandableListView expandLV, int groupPosition, int
-                Uri taskUri = ContentUris.withAppendedId(Tasks.getContentUri(mAuthority), selectTaskId);
+            // TODO For now we get the id of the task, not the instance, once we support recurrence we'll have to change that, use instance URI that time
+            Uri taskUri = ContentUris.withAppendedId(Tasks.getContentUri(mAuthority), (long) TaskFieldAdapters.TASK_ID.get(cursor));
+            // TODO Should we use TASK_COLOR? (That's not in the projection.)
```
No, `TASK_COLOR` should be used as some sort of secondary color, like on a banner, badge or additional color bar.

lemonboston replied:
Thanks, good to know. I've updated the parameter names to refer to the task list color.

lemonboston commented:
@dmfs I've made the color args mandatory for the Fragments as discussed on Slack. It allowed to remove classes, simplify the code.

I've kept `Conditional` and `OptionalNonTransparentIntColor` they seem okay for me, but that usage could be replaced with a ternary operator, let me know if you would prefer that.

lemonboston commented:
I've changed my mind about those 2 classes, not worth adding and maintaining them for a single `if`, so I've removed them.

dmfs commented:
On tablets it looks a bit strange that the details view is faded while the main app bar just changes instantly. For now it's ok but we should think of something to improve this later on.

dmfs commented:
I like that the app bar stays with the color of the last task when you delete it, instead of going back to the primary color.

dmfs commented:
Rotating on a small tablet still doesn't work properly.

## Steps to reproduce:

* open a task in portrait mode, the details should show up full screen
* rotate to landscape

Expected result:

* same task is opened in the details pane

Actual result:

* a different task is opened in the details pane

## Similarly (and probably related)

* select a task in landscape mode
* rotate to portrait (the list view shows up)
* rotate back to landscape

Expected result:

* the same task is selected 

Actual result:

*  a different task is selected

dmfs commented on opentasks/src/main/java/org/dmfs/tasks/ViewTaskActivity.java:
```diff
@@ -65,7 +67,7 @@ protected void onCreate(Bundle savedInstanceState)
         if (savedInstanceState == null)
         {
-            ViewTaskFragment fragment = ViewTaskFragment.newInstance(getIntent().getData());
+            ViewTaskFragment fragment = ViewTaskFragment.newInstance(getIntent().getData(), new PrimaryColor(this));
```
How about passing this color to the activity as well? On my device I see the primary color briefly when opening a task, we could avoid that by passing the color to the `ViewTaskActivity`. Just make sure you tolerate the absence of the color in case it's called by an external app.

lemonboston replied:
I've added color as an optional extra to `ViewTaskActivity`.

lemonboston commented:
Strange, I couldn't reproduce the first case you described, that works as expected for me. I had also tested this.
And for the second case, for me the empty fragment is shown when turning back to landscape. With correct color, so that is saved, but the selected task is not.

I compared with master, I see it works correctly there, and that the selected task item is highlighted in portrait mode as well.

I am going to look into this.

dmfs commented:
Make sure you create multiple tasks and select one of the latter ones.

lemonboston commented:
I've pushed a branch named `mx/624-investigation` which add some logs on top of this branch.
Here are the logs from the Nexus 7 Android 8.0 emulator, for the following action:

- clean install, 3 tasks created with quick add dialog
- select the last one (the log starts here)
- rotate once
- rotate again

```
01-16 17:48:26.723 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@4b9ca7a.onItemSelected() | uri = content://org.dmfs.tasks/tasks/3 | color = -15042873 | force = true | pos = 0
                                                                at org.dmfs.tasks.TaskListFragment.selectChildView(TaskListFragment.java:513)
                                                                at org.dmfs.tasks.TaskListFragment.access$000(TaskListFragment.java:87)
                                                                at org.dmfs.tasks.TaskListFragment$1.onChildClick(TaskListFragment.java:154)
                                                                at android.widget.ExpandableListView.handleItemClick(ExpandableListView.java:714)
                                                                at android.widget.ExpandableListView.performItemClick(ExpandableListView.java:653)
01-16 17:48:26.763 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@4b9ca7a.replaceTaskDetailsFragment() | fragment = ViewTaskFragment{184c577}
                                                                at org.dmfs.tasks.TaskListActivity.onItemSelected(TaskListActivity.java:410)
                                                                at org.dmfs.tasks.TaskListFragment.selectChildView(TaskListFragment.java:513)
                                                                at org.dmfs.tasks.TaskListFragment.access$000(TaskListFragment.java:87)
                                                                at org.dmfs.tasks.TaskListFragment$1.onChildClick(TaskListFragment.java:154)
01-16 17:48:53.911 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@9999c9.onCreate() | selectedUri = null | color = -15042873 | savedState = Bundle[{android:viewHierarchyState=Bundle[{android:views={16908290=android.view.AbsSavedState$1@ac589f4, 2131296261=android.view.View$BaseSavedState@cdc451d, 2131296272=android.view.View$BaseSavedState@8467392, 2131296284=android.view.View$BaseSavedState@221e163, 2131296352=android.view.AbsSavedState$1@ac589f4, 2131296398=FragmentPager.SavedState{ab5160 position=0}, 2131296452=HorizontalScrollView.SavedState{bfd1619 scrollPosition=0}, 2131296455=android.view.View$BaseSavedState@7d436de, 2131296510=android.support.v7.widget.Toolbar$SavedState@bc730bf}, android:focusedViewId=16908298}], android:support:fragments=android.support.v4.app.FragmentManagerState@764238c, mCurrentPageId=0, mLastUsedColor=-15042873, @android:autofillResetNeeded=true, mShouldShowDetails=true, mShouldSwitchToDetail=false, mSelectedTaskUri=null, android:lastAutofillId=1073741874, android:sessionId=623073871}]
01-16 17:48:54.600 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@9999c9.onItemSelected() | uri = content://org.dmfs.tasks/tasks/3 | color = -15042873 | force = false | pos = 0
                                                                at org.dmfs.tasks.TaskListFragment.selectChildView(TaskListFragment.java:513)
                                                                at org.dmfs.tasks.TaskListFragment.access$000(TaskListFragment.java:87)
                                                                at org.dmfs.tasks.TaskListFragment$7.run(TaskListFragment.java:911)
                                                                at android.os.Handler.handleCallback(Handler.java:789)
                                                                at android.os.Handler.dispatchMessage(Handler.java:98)
01-16 17:49:24.418 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@8b64b62.onCreate() | selectedUri = null | color = -15042873 | savedState = Bundle[{android:viewHierarchyState=Bundle[{android:views={16908290=android.view.AbsSavedState$1@ac589f4, 2131296261=android.view.View$BaseSavedState@c4a481a, 2131296272=android.view.View$BaseSavedState@bb4cf4b, 2131296352=android.view.AbsSavedState$1@ac589f4, 2131296398=FragmentPager.SavedState{2274d28 position=0}, 2131296452=HorizontalScrollView.SavedState{92e3a41 scrollPosition=0}, 2131296510=android.support.v7.widget.Toolbar$SavedState@db4cbe6}, android:focusedViewId=16908298}], android:support:fragments=android.support.v4.app.FragmentManagerState@5ff1527, mCurrentPageId=0, mLastUsedColor=-15042873, @android:autofillResetNeeded=true, mShouldShowDetails=false, mShouldSwitchToDetail=false, mSelectedTaskUri=null, android:lastAutofillId=1073741874}]
01-16 17:49:24.419 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@8b64b62.replaceTaskDetailsFragment() | fragment = EmptyTaskFragment{93813d4}
                                                                at org.dmfs.tasks.TaskListActivity.onCreate(TaskListActivity.java:234)
                                                                at android.app.Activity.performCreate(Activity.java:6975)
                                                                at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1213)
                                                                at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2770)
01-16 17:49:25.040 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@8b64b62.onItemSelected() | uri = content://org.dmfs.tasks/tasks/3 | color = -15042873 | force = false | pos = 0
                                                                at org.dmfs.tasks.TaskListFragment.selectChildView(TaskListFragment.java:513)
                                                                at org.dmfs.tasks.TaskListFragment.access$000(TaskListFragment.java:87)
                                                                at org.dmfs.tasks.TaskListFragment$7.run(TaskListFragment.java:911)
                                                                at android.os.Handler.handleCallback(Handler.java:789)
                                                                at android.os.Handler.dispatchMessage(Handler.java:98)
01-16 17:49:25.043 7738-7738/org.dmfs.tasks D/Investigator: [main] TaskListActivity@8b64b62.replaceTaskDetailsFragment() | fragment = ViewTaskFragment{ff03f0b}
                                                                at org.dmfs.tasks.TaskListActivity.onItemSelected(TaskListActivity.java:410)
                                                                at org.dmfs.tasks.TaskListFragment.selectChildView(TaskListFragment.java:513)
                                                                at org.dmfs.tasks.TaskListFragment.access$000(TaskListFragment.java:87)
                                                                at org.dmfs.tasks.TaskListFragment$7.run(TaskListFragment.java:911)
```

lemonboston commented:
I've found the bug for the task list item selection highlight - `TaskListFragment` had the `twoPane` variable as fragment argument, so it kept its value after rotation. And the highlighting is based on that.
I've changed it to simply read the `has_two_pane` bool resource which I suppose we should use at all places.
Ready for checking again.

lemonboston commented:
I've added two more changes today, to fix issues on small tablet:
-- fix the case when selected item in landscape mode is deleted by swiping it out - it now replaces the fragment with EmptyTaskFragment and the list is shown after rotation
-- fix the case of missing selection highlight when rotating to landscape, also fixing to keep the correct task selected

Note that some of the issues we encountered are alreay present in master version.
I've added bugs: #640, #641, #642

#642 should be partially fixed here.

The behaviour for showing the details in portrait mode is still not perfect:
it works if we don't change selection in landscape mode, but if we do, then the list is shown after rotating back to portrait.

But as I saw on master, this doesn't work at all there, even without changing selection.

So for this case that still doesn't work, we could add a low priority bug.

Ready for check again.

dmfs commented:
It's expected behavior that the list is shown when you select a new task in landscape mode and rotate to portrait mode.

dmfs commented:
The latest version crashes for me when adding a task in single pane mode:
```
01-18 22:03:52.175  4629  4629 E AndroidRuntime: Process: org.dmfs.tasks, PID: 4629
01-18 22:03:52.175  4629  4629 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.Integer.intValue()' on a null object reference
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at org.dmfs.tasks.model.adapters.ColorFieldAdapter.get(ColorFieldAdapter.java:93)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at org.dmfs.tasks.EditTaskFragment.saveAndExit(EditTaskFragment.java:798)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at org.dmfs.tasks.EditTaskFragment.onOptionsItemSelected(EditTaskFragment.java:539)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v4.app.Fragment.performOptionsItemSelected(Fragment.java:2379)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v4.app.FragmentManagerImpl.dispatchOptionsItemSelected(FragmentManager.java:3133)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v4.app.FragmentController.dispatchOptionsItemSelected(FragmentController.java:344)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v4.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:414)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:195)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:113)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.app.AppCompatDelegateImplV9.onMenuItemSelected(AppCompatDelegateImplV9.java:679)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:822)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:156)
01-18 22:03:52.175  4629  4629 E AndroidRuntime: 	at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:969)
```

dmfs commented:
crashes in two-pane mode as well

dmfs commented on opentasks/src/main/java/org/dmfs/tasks/EditTaskFragment.java:
```diff
@@ -793,7 +793,9 @@ public void saveAndExit()
-                    activity.startActivity(new Intent("android.intent.action.VIEW", mTaskUri));
+                    activity.startActivity(
+                            new Intent(Intent.ACTION_VIEW, mTaskUri)
+                                    .putExtra(ViewTaskActivity.EXTRA_COLOR, TaskFieldAdapters.LIST_COLOR.get(mValues)));
```
This won't work and causes the aforementioned crash because `mValues` doesn't contain the list color at this point. If the task is new, `mValues` has been created from scratch and not been loaded from the database, so it doesn't contain the list color yet.

dmfs replied:
There is a field called `mListColor` which holds the color of the currently selected list.

lemonboston commented:
I've fixed the first crash about the color.

I couldn't reproduce the 2nd one when deleting a task. I tried to figure it out from the stacktrace what may cause that and applied a speculative fix.
If you remember or can reproduce it, could you please describe the steps?

Ready for check again.

dmfs commented:
This is still not working correctly. Try on a phone:
* open task editor
* enter title
* save
* press back to return to the list
* rotate the device

### Expected behavior
The list is still shown

### Actual behavior
The just closed details view is shown

dmfs commented:
I'll look into that.
